### ISSUE TITLE
Crash when accessing vlen of strings

### ISSUE THREAD

**Issue by gavinmacaulay** [h5py/h5py#1336]
h5py    2.9.0
HDF5    1.10.4
Python  3.6.9 |Anaconda, Inc.| (default, Jul 30 2019, 14:00:49) [MSC v.1915 64 bit (AMD64)]
sys.platform    win32
sys.maxsize     9223372036854775807
numpy   1.16.4

I get a crash when trying to access a vlen of strings variable. I have a minimal example, defined via a netcdf4 .cdl file:

```
netcdf error {
	types:
		string(*) name_t;
	dimensions:
		n = 2;
	variables:				
		name_t names(n);
	data:
		names = {"name1.1", "name1.2"}, {"name2"};
}
```
(I can provide the resulting hdf5 file if necessary) 

Which when accessed via this code:
```
import h5py
with h5py.File('error.nc', 'r') as f:
    names = f['names']
    for n in names:
        print(n)       
```

crashes the python interpreter on the print(n) line. I think that this is probably the same problem as reported on the h5py google groups list:  https://groups.google.com/forum/#!topic/h5py/mi3ndCigBEA

The example file is read fine by other hdf5 readers (h5dump, HDFView, Matlab, etc).

**Comment by takluyver** [comment]
I agree it's similar to the mailing list post from last month. Possibly also to #623. The key appears to be having a datatype which has vlen arrays of vlen strings. I've simplified an example from the mailing list thread - this reproduces the issue for me:

```python
import h5py
import numpy as np

# Create file with dataset containing vlen arrays of vlen strings
with h5py.File("h5py_string_test2.h5", "w") as h:
    vlen_str = h5py.special_dtype(vlen=str)
    vlen_vlen_str = h5py.special_dtype(vlen=vlen_str)

    ds = h.create_dataset('/com', (2,), dtype=vlen_vlen_str)
    ds[0] = (np.array(["a", "b", "c"], dtype=vlen_vlen_str))
    ds[1] = (np.array(["d", "e", "f","g"], dtype=vlen_vlen_str))

# At this point, h5py_string_test2.h5 appears to be a valid
# file, i.e. can be h5dump'd, and the contents look fine

# Read the file
with h5py.File("h5py_string_test2.h5", "r") as h:
    ds = h["com"]
    print("Shape of ds: ", ds.shape)
    print("Dtype:", ds.dtype)
    print("Dtype metadata:", ds.dtype.metadata)
    print("vlen sub-dtype metadata:", ds.dtype.metadata['vlen'].metadata)

    print(ds[0])  # segfaults
```

**Comment by takluyver** [comment]
I think it's also worth saying that what HDF5 does well is storing regular arrays of numbers. This is absolutely a bug, and I'm not saying it won't get fixed, but if you want to store varying numbers of variable-length strings, it might be worth considering whether another storage format fits your needs better.

**Comment by gavinmacaulay** [comment]
Indeed. The strings I want to store are a small part of a file that contains mostly rectangular arrays of numbers. For the moment, I will avoid the problem by storing the strings in a vector and then have a matching numeric id vector to give the grouping of the strings.

**Comment by egachen** [comment]
I got the same problem too, can you try this fix? https://github.com/h5py/h5py/pull/1393
